Add documentation - Getting Started : Access Control Rights (#1813)
kuzzleio/kuzzle: doc/2/guides/getting-started/3-access-control-rights/index.md

title: Access Control Rights
description: Define user rights
order: 300
---

# Access Control Rights

> As in any backend, Kuzzle allows you to restrict access to its functionalities according to the users.

Reviewer: ```suggestion
As in any backend, Kuzzle allows you to restrict access to its features, depending on the querying users.
```

The permissions system is designed following a standard model and is structured in 3 dimensions:

>  - **role**: whitelist authorized API actions

Reviewer: ```suggestion
 - **role**: whitelist of allowed API actions
```

 - **profile**: combination of one or more roles
 - **user**: combination of one or more profiles

![roles, profiles and users](./role-profile-user.png)

See the [in-depth guide](/core/2/guides/some-link-for-acl-in-depth) for more details.

### Create our first restricted user

#### Role

> First, we are gonna create a new role with the [security:createRole](/core/2/api/controllers/security/create-role) action.

Reviewer: ```suggestion
First, we are going to create a new role with the [security:createRole](/core/2/api/controllers/security/create-role) action.
```

> The following role description give access to [auth:getCurrentUser](/core/2/api/controllers/auth/get-current-user) and [server:info](/core/2/api/controllers/auth/get-current-user) actions only.

Reviewer: ```suggestion
The following role description gives access to [auth:getCurrentUser](/core/2/api/controllers/auth/get-current-user) and to the [server:info](/core/2/api/controllers/auth/get-current-user) actions only.
```

```js
{
  controllers: {
    auth: {
      actions: {
        getCurrentUser: true
      }
    },
    server: {
      actions: {
        now: true
      }
    }
  }
}
```

> Copy the above role and then run this command: `kourou security:createRole --id hostes --body-editor`

Reviewer: Why that id? I don't get it :thinking:

> A text editor should open itself, replace the existing body by our role definition and then exit the editor.

Reviewer: ```suggestion
A text editor should open itself: replace the existing body by the copied role definition and then exit the editor.
```

::: info
If you have trouble with the integrated text editor, you can try to run the command directly as well:  

```bash
$ kourou security:createRole '{
  controllers: {
    auth: {
      actions: {
        getCurrentUser: true
      }
    },
    server: {
      actions: {
        now: true
      }
    }
  }
}' --id hostes
```
:::

You should see your newly created role in the `Security > Roles` section of the [Admin Console](http://console.kuzzle.io)

![Admin Console roles display](./admin-console-roles.png)

#### Profile

> Then, we are gonna create a profile who use our newly created role. For this we will use the [security:createProfile](/core/2/api/controllers/security/create-profile) action.

Reviewer: ```suggestion
Then, we are going to create a profile which uses our newly created role. For this we will use the [security:createProfile](/core/2/api/controllers/security/create-profile) action.
```

```bash
$ kourou security:createProfile '{
  policies: [
    { roleId: "hostes" }
  ]
}' --id hostes
```

> Now we have a `hostes` profile which give access to the `hostes` role actions.

Reviewer: ```suggestion
Now we have a `hostes` profile which gives access to the API actions allowed by the `hostes` role.
```

You should see your newly created profile in the `Security > Profiles` section of the [Admin Console](http://console.kuzzle.io)

![Admin Console profiles display](./admin-console-profiles.png)

#### User

> Finally, we need an user that use the `hostes` profile. The API action to create an user is [security:createUser](/core/2/api/controllers/security/create-user).

Reviewer: ```suggestion
Finally, we need a user attached the `hostes` profile. The API action to create a user is [security:createUser](/core/2/api/controllers/security/create-user).
```

> User need at least to be assigned with one profile. Also we are gonna give our user some credentials to be able to log in.

Reviewer: ```suggestion
Users need to have at least one assigned profile. We also will have to give our user some credentials to be able to log in with it.
```

```bash
$ kourou security:createUser '{
  content: {
    profileIds: ["hostes"]
  },
  credentials: {
    local: {
      username: "yagmur",
      password: "password"
    }
  }
}'
```

You should see your newly created role in the `Security > Users` section of the [Admin Console](http://console.kuzzle.io)

![Admin Console users display](./admin-console-users.png)

> ### Creates administrator and restrict anonymous user rights

Reviewer: ```suggestion
### Creating an administrator account, and restricting anonymous user rights
```

> When you are not authenticated, your request are executed as the `anonymous` user.

Reviewer: ```suggestion
When you are not authenticated, your requests are executed as the `anonymous` user.
```

> Like every other user, this user is assigned with a profile (`anonymous`) and thus a role (`anonymous` as well).

Reviewer: ```suggestion
As with any other user, the `anonymous` user has a profile assigned (named `anonymous`), and thus a role (named `anonymous` as well).
```

::: info

> By default, the `anonymous` role give access to every API actions. This is intended to makes development easier but it's definitively not suitable for production.

Reviewer: ```suggestion
By default, the `anonymous` role gives access to all API actions. This is intended to make development easier, but it's definitively not suitable for production.
```

:::

> It's recommanded to use the [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) action to creates an admin and restrict anonymous user rights.

Reviewer: ```suggestion
It's recommended to use the [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) action to create an administrator user, and to restrict anonymous user rights.
```

::: info

> The [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) creates an user which have the `admin` profile and this the `admin` role who give access to every API actions.

Reviewer: ```suggestion
The [security:createFirstAdmin](/core/2/api/controllers/security/create-first-admin) creates a user attached to the `admin` profile, which uses the `admin` role, giving access to all API actions.
```

:::

> In this way you can always access the complete API throught this admin account.

Reviewer: ```suggestion
This way you can always access the complete API through this admin account.
```

```bash
$ kourou security:createFirstAdmin '{
  credentials: {
    local: {
      username: "admin",
      password: "password"
    }
  }
}' -a reset=true
```

> #### Try the API as anonymous user

Reviewer: ```suggestion
#### Try the API as the anonymous user
```

Try to run the following command: `kourou server:now`

> You should get the following error because since the anonymous user is now restricted to only a few API actions:

Reviewer: ```suggestion
You should get the following error because now the anonymous user is restricted to only a few API actions:
```

```bash
$ kourou server:now

[ℹ] Unknown command "server:now", fallback to API method
 
 🚀 Kourou - Executes an API query.
 
 [ℹ] Connecting to http://localhost:7512 ...
 [X] UnauthorizedError: Unauthorized: authentication required to execute the action "server:now". -1
    [...Kuzzle internal calls deleted...]
    at Funnel.checkRights (/var/app/lib/api/funnel.js:374:28)
        status: 401
        id: security.rights.unauthorized
```

> #### Try the API as authentified user

Reviewer: ```suggestion
#### Try the API as an authenticated user
```

> Run the same command but authenticate with the user we just create: `kourou server:now --username yagmur --password password`

Reviewer: ```suggestion
Run the same command, authenticating with the user we just created: `kourou server:now --username yagmur --password password`
```

> We have the right to use this API action because we are now authenticated.

Reviewer: ```suggestion
We are allowed to use this API action because we are now authenticated with a user with sufficient rights.
```
